### PR TITLE
Replace ushort usage with int

### PR DESCRIPTION
**Pull request by MihaZupan**
Uri code wastes cpu cycles by using ushort for indexes and lengths. This PR changes those to int.

Also removed `FindEndOfComponent`, as it was effectively a glorified and confusing `IndexOf`.

### CODE REVIEW

**Comment by danmosemsft** [comment]
Interesting, how did you determine this was a performance issue? Just trying the change? How much difference does it make?

**Comment by scalablecory** [comment]
I had assumed it was using ushort because signed integers can be a little slower. When indexing into an array, you see an extra `movsxd`:

```
movsxd rdx, edx
mov rax, [rcx + rdx]
```

Though `uint` may make more sense than `ushort`. I believe on x64 most 16-bit instructions require an extra prefix that will bloat your code size.

**Comment by stephentoub** [comment]
Any throughput-focused change like this should have performance numbers accompanying it to demonstrate the value in the churn.

**Comment by MihaZupan** [comment]
@scalablecory using ushort for indexing means you end up having to extend the 16bit value on every loop iteration.

You can see the overhead on a simple method like
```c#
char Get(char* source, ushort index) => source[index];
```

```asm
Get(Char*, UInt16)
    L0000: movzx eax, dx
    L0003: movsxd rax, eax
    L0006: movzx eax, word [rcx+rax*2]
    L000a: ret

Get(Char*, Int32)
    L0000: movsxd rax, edx
    L0003: movzx eax, word [rcx+rax*2]
    L0007: ret
```

[More examples](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAZgAJiAmcgYXIG9znyBYAKBcouIEYlyAVwB2ubADMYQ3AAtoGcgFVZ8gBRgZ2KACpyuKGDTS5UBQBsYwgOYYZRjVvJgAlBy4M3XFuOjlVglVNyAEtyAF5yAAYAbhDyAB5yC2tbWOCAanTXTi9mDxzcrmDxP30wAG1ggF1wiJdPQsLiAHYQ6IavAF8OrhakyxsZdoLybpGySn4hUQkpYOEFAEkF9U0dPQMjefMB23s1p2z3Hu9fVW24iJi4xOTBtMyjwvzGopLVMsqasLqn168+sFho0xo0+ndUh1QSwOhM+AIRGJJMZ5EpAhgqKstLoykYAiYdik7E4DvURi9CvjUaErsDCgB3GTBCx+UK3XYycgAMi5Gwq1XIAEJfiFMnTcoDxaMOrCeFNEbMQgtyMsMVj1rilYTBvtHGTjiMisqaVEpVxGcypOcEv0idzeZ8BcLDqL0maWJKoTLxnKETNkVSgsoCaRVHxIny8eiQgATBBRgm2nUkvV/RgnZiBhSWGPhPxZ5znOPkdJJ2zOd2Z6Mm4JxyvkC0s62JHP2vlfIUijJujOUVpAr0jWWTP1IubK1Wh8ORrWx+OziHEhxQQ4dCm5C6tiK1hAlstDXsXGt13uNq1s8itnntp1dsW9z0jaHMYfw6ZjlFB9EoMO8COarM5wTVFF11Fd9RYdcvEA48EHrM9WRtfx0ULHc90XZw20dGpnRcV160fLhnzYH0R3fRULlVH9p01I84y2ZVQJTcC0ygo0FFg+CmSbC80NLRcsIMDtcMw7sCP7KViNfeV/SkYgUDRAlqL/GdYBKGD6M/bNhBjMCnF4PSwCoVjexgvMdy4y1EJbHTBP5HC6l4OzhLqYz8MPYtt0k71el9cjkXklUFmU/9NnINTZx3BjtN05j9MM4y10PY1zJPQ0WAQ5tL1s69sM7fTnNvJw3LEjzdy8wcuGk0dFUC4N5AAVnVHEwo0+cIsAnNDJM9LmEytCbNzXKhJ3ByXR3e9euy3NtzSoifI9PyFQChTVSa5cWsMSLNIizcdO6pKpv64tBsKuMxrwiaeymrc528odSLfZdFPkJBf3/YCgh3TCwgAPj0Spzqlaq4tVN6aOiucfv+3BAYQKpgcemSP2e+rTGaZq9E+ji42hgHRsR3yyOWqRntVDGNqx7aEDx2GCYW5g4WRxVgAgCAzBe0wAA5MYA6tNOeiDcj+m9ztqJxCcW4nZPIVn2aCjAecp2jjQF0k0y4EXHTFn4JY4TogA) of the same.

I don't think there is ever an upside for using ushort over 32/64bit types when iterating.

> Interesting, how did you determine this was a performance issue? Just trying the change? How much difference does it make?

I was surprised to see ushort being used in this way when first looking at Uri code. I ran a quick microbenchmark comparing between int16, uint16, int32, uint32, int64, uint64 for such cases just to make sure there isn't some ushort magic I'm missing out on.

Uri uses ushort to store indexes in a packed struct to reduce the size of a Uri allocation and that makes sense (this is also the source of the Uri length limitations).
But it also takes steps to ensure that it uses ushort for indexes, lengths in a lot of places, and in such scenarios it doesn't make sense.

The perf diff of this change is anywhere from 0 to ~10% depending on the code paths hit. For the simple scenario of doing `new Uri("https://www.microsoft.com/")`, the difference is ~5-10%.

**Comment by scalablecory** [comment]
Thanks. Do you see any perf change with `uint`? It will get rid of that stray `movsxd` vs the `int` version.

File: src/libraries/System.Private.Uri/src/System/UncNameHelper.cs [hunk 1]
```diff
@@ -49,34 +49,35 @@ internal static string ParseCanonicalName(string str, int start, int end, ref bo
...
                     ++i;
                     break;
                 }
-                if (char.IsLetter(name[i]) || name[i] == '-' || name[i] == '_')
+                if (char.IsLetter(c) || c == '-' || c == '_')
                 {
                     validShortName = true;
                 }
-                else if (name[i] < '0' || name[i] > '9')
+                else if (c < '0' || c > '9')
```
Comment (EgorBo) [inline src/libraries/System.Private.Uri/src/System/UncNameHelper.cs]: `c - '0' > '9' - '0'` ?

Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/UncNameHelper.cs]: Saw this as well, I am planning on changing similar checks as a part of a different PR in the uri-cleanup series.

Comment (EgorBo) [reply src/libraries/System.Private.Uri/src/System/UncNameHelper.cs]: I think we need to make `char.IsInRange` public for that. The hack I suggested is widely used in the BCL but looks ugly. I had a draft PR to JIT to handle this pattern there https://github.com/dotnet/coreclr/pull/27480

Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/UncNameHelper.cs]: I was thinking of isolating common character checks into a Uri internal `CharHelper` class like
```c#
internal static class CharHelper
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiDigit(char c) =>
        (uint)(c - '0') <= ('9' - '0');

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiLowercaseLetter(char c) =>
        (uint)(c - 'a') <= ('z' - 'a');

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiUppercaseLetter(char c) =>
        (uint)(c - 'A') <= ('Z' - 'A');

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiLetter(char c) =>
        (uint)((c - 'A') & ~0x20) <= ('Z' - 'A');

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiLetterOrDigit(char c) =>
        ((uint)((c - 'A') & ~0x20) <= ('Z' - 'A')) ||
        ((uint)(c - '0') <= ('9' - '0'));

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsHexDigit(char c) =>
        ((uint)(c - '0') <= ('9' - '0')) ||
        ((uint)((c - 'A') & ~0x20) <= ('F' - 'A'));

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsInInclusiveRange(char c, char min, char max) =>
        (uint)(c - min) <= (uint)(max - min);
}
```

**Comment by alnikola** [comment]
> The perf diff of this change is anywhere from 0 to ~10% depending on the code paths hit. For the simple scenario of doing `new Uri("https://www.microsoft.com/")`, the difference is ~5% (re-running the benchmark multiple times as it yields results from 0 to 15%, averaging at ~5).

@MihaZupan 0% - 15% fluctuations looks suspiciously large. Could you please post the statistical metrics calculated by BenchmarkDotNet for the original and changed versions? Especially, Mean and Standard Deviation.

**Comment by MihaZupan** [comment]
@alnikola  For reference, [here's the dump](https://gist.github.com/MihaZupan/f2ec812325c55520655138346b4ca756) of 40 benchmark runs, ran on the work laptop for
```c#
[Benchmark] public Uri NewUri() => new Uri("https://www.microsoft.com/");
```
The average of those runs is a 3.3% improvement.

I will post results of the runs from a desktop machine when I get home.

@adamsitnik When running BenchmarkDotNet from the command line with `--coreRun`s specified, is there a way to specify the job type (for example LongRunJob)? Or can the coreRuns be specified in the ManualConfig?

**Comment by adamsitnik** [comment]
> When running BenchmarkDotNet from the command line with --coreRuns specified, is there a way to specify the job type (for example LongRunJob)? Or can the coreRuns be specified in the ManualConfig?

The following command should work:

```cmd
--job LongRunJob --corerun $path
```

**Comment by MihaZupan** [comment]
And a verylong run on the same machine yields
| Method |          Toolchain |     Mean |   Error |  StdDev |   Median | Ratio | RatioSD |
|------- |------------------- |---------:|--------:|--------:|---------:|------:|--------:|
| NewUri | \clean\CoreRun.exe | 199.3 ns | 0.59 ns | 7.89 ns | 198.0 ns |  1.04 |    0.04 |
| NewUri |   \new\CoreRun.exe | 191.6 ns | 0.17 ns | 2.20 ns | 192.4 ns |  1.00 |    0.00 |

**Comment by alnikola** [comment]
> And a verylong run on the same machine yields

This looks fine to me.

**Review by alnikola — APPROVED** [review]
LGTM

**Review by wfurt — APPROVED** [review]
LGTM

**Comment by jaykrell** [comment]
16bit integers should generally be avoided for perf as I understand. Except for storage size.

**Comment by stephentoub** [comment]
Are there any places where we may have been eliminating bounds checks, e.g. because the JIT could prove that the value was < the length (and being unsigned it definitely wasn't < 0), and now with a signed int it may no longer be able to eliminate those bounds checks?

**Comment by EgorBo** [comment]
> Are there any places where we may have been eliminating bounds checks, e.g. because the JIT could prove that the value was < the length (and being unsigned it definitely wasn't < 0), and now with a signed int it may no longer be able to eliminate those bounds checks?

In theory it's possible 🙂 https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC26A3tmUWuRVVrgCzYCySAFAJYB2wA2gLrZgAToLABPADTYArhAAWAe0HBsbAJToqxDVXJsAZthYspnYKqEjRAOgAyAUw4BzYLOwA+Feso6fFsdzZ+AF5sAAYAbm0yAF8o6nJaOAZGBHYuPgFhMUlTT2wAenzsAGNZMCc7ABNsYHkVLjitbx99Q2NTcyzreycXd084nzI/UQDgsMjm7Fi0aKA=

**Comment by stephentoub** [comment]
> In theory it's possible 

Exactly.  Hence my question :wink:  Thanks for the concrete example.

**Comment by MihaZupan** [comment]
I will recheck and [fix if necessary](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC26A3tmUWuRVVrgCzYCySAFAJYB2wA2gLrZgAToLABPADTYArhAAWAe0HBsbAJToqxDVXJsAZthYspnYKqEjRAOgAyAUw4BzYLOwA+Feso6fFsdzZ+AF5sAAYAbm0yAF8o6nJaOAZGBHYuPgFhMUlTT2wAenzsAGNZMCc7ABNsYHkVLjitbx99Q2NTcyzreycXd084nzI/UQDgsMjm7FipuMTkzDSefhGcrgGppqGVAyMTLlU2bAAeQ32zEdsHZ1kvbaoRsewQiLiZ6KA=), tho I haven't seen any explicit casts of `string.Length` in the code, so there might not be any regressions related to it, but we can definitely improve it in a few places.

**Comment by MihaZupan** [comment]
>Are there any places where we may have been eliminating bounds checks, e.g. because the JIT could prove that the value was < the length (and being unsigned it definitely wasn't < 0), and now with a signed int it may no longer be able to eliminate those bounds checks?

I don't see any such cases

**Comment by stephentoub** [comment]
@MihaZupan, should this be merged?

**Pull request merged**